pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -723,6 +917,115 @@ void topicTerminated() {
+                }
+                marketDeleteFuture.complete(null);
+            }
+
+            @Override
+            public void markDeleteFailed(ManagedLedgerException exception, Object ctx) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}][{}] Failed to mark delete for position ", topicName, subName, ctx, exception);
+                }
+                marketDeleteFuture.completeExceptionally(exception);
+            }
+        };
+
+        DeleteCallback deleteCallback = new DeleteCallback() {
+            @Override
+            public void deleteComplete(Object position) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}][{}] Deleted message at {}", topicName, subName, position);
+                }
+                deleteFuture.complete(null);
+            }
+
+            @Override
+            public void deleteFailed(ManagedLedgerException exception, Object ctx) {
+                if (log.isDebugEnabled()) {
+                    log.warn("[{}][{}] Failed to delete message at {}", topicName, subName, ctx, exception);
+                }
+                deleteFuture.completeExceptionally(exception);
+            }
+        };
+
+        // It's valid to create transaction then commit without doing any operation, which will cause
+        // pendingAckMessagesMap to be null.
+        ConcurrentOpenHashSet<Position> pendingAckMessageForCurrentTxn = pendingAckMessagesMap != null ?
+                                                this.pendingAckMessagesMap.remove(txnId) : new ConcurrentOpenHashSet();
+        List<Position> positions = pendingAckMessageForCurrentTxn.values();
+        // Materialize all single acks.
+        cursor.asyncDelete(positions, deleteCallback, positions);
+        if (pendingAckMessages != null) {
+            positions.forEach(position -> this.pendingAckMessages.remove(position));
+        }
+
+        // Materialize cumulative ack.
+        cursor.asyncMarkDelete(this.pendingCumulativeAckMessage, (null == properties)?
+                Collections.emptyMap() : properties, markDeleteCallback, this.pendingCumulativeAckMessage);
+
+        // Reset txdID and position for cumulative ack.
+        PENDING_CUMULATIVE_ACK_TXNID_UPDATER.set(this, null);
+        POSITION_UPDATER.set(this, null);
+        deleteFuture.runAfterBoth(marketDeleteFuture, () -> commitFuture.complete(null))
+                    .exceptionally((exception) -> {
+                        commitFuture.completeExceptionally(exception);
+                        return null;
+                    });
+
+        return commitFuture;
+    }
+
+    /**
+     * Abort a transaction.
+     *
+     * @param txnId  {@link TxnID} to identify the transaction.
+     * @throws IllegalArgumentException if given {@link TxnID} is not found in this subscription.
+     */
+    public synchronized CompletableFuture<Void> abortTxn(TxnID txnId) {
```

Comment:
is abortTxn need to redeliver messages the txnId contains? because if broker handle 
 redeliverUnacknowledgedMessages will skip the messages which is in a transaction, if abortTxn do not redeliver these messages, i think message will never be redelivery.

Reply:
As discussed, redelivery all un acked message after transaction abort is more desired behavior for most transaction use case. So will update to call dispatcher.redeliverUnacknowledgedMessages after abort a transaction.